datafusion/physical-expr/src/equivalence/properties.rs

```diff
@@ -1828,22 +1832,32 @@ fn calculate_union_binary(
     }
 
     // First, calculate valid constants for the union. An expression is constant
-    // at the output of the union if it is constant in both sides.
-    let constants: Vec<_> = lhs
+    // at the output of the union if it is constant in both sides with matching values.
+    let constants = lhs
         .constants()
         .iter()
-        .filter(|const_expr| const_exprs_contains(rhs.constants(), const_expr.expr()))
-        .map(|const_expr| {
-            // TODO: When both sides have a constant column, and the actual
-            // constant value is the same, then the output properties could
-            // reflect the constant is valid across all partitions. However we
-            // don't track the actual value that the ConstExpr takes on, so we
-            // can't determine that yet
-            ConstExpr::new(Arc::clone(const_expr.expr())).with_across_partitions(false)
+        .filter_map(|lhs_const| {
+            // Find matching constant expression in RHS
+            rhs.constants()
+                .iter()
+                .find(|rhs_const| rhs_const.expr().eq(lhs_const.expr()))
+                .map(|rhs_const| {
+                    // Create new constant with across_partitions=false since we're in a union
+                    let mut const_expr = ConstExpr::new(Arc::clone(lhs_const.expr()))
+                        .with_across_partitions(false);
```

Comment:
Rather than directly setting `across_partitions` flag as false, can we do a better estimation by checking the `lhs` and `rhs` values? If they are the same, then there should be no harm to set this flag as true?

Reply:
Resolved in de8bc13.